File: src/main/java/org/processmining/utils/helpers/BaseGenerationHelper.java
package org.processmining.utils.helpers;

import org.processmining.models.*;

import java.util.*;

/**
 * Created by Ivan Shugurov on 23.10.2014.
 */

/**
 * @param <T> type of objects which can contain tokens, for example, places in Petri nets
 * @param <K> type of objects it's possible to move through. For example, transitions in Petri nets
 * @param <F> type of additional movables possible during a replay
 */
public abstract class BaseGenerationHelper<T extends Tokenable, K extends Movable, F extends Movable> implements GenerationHelper<K, F>
{
    protected static final Random random = new Random();
    private final Set<T> initialMarking;
    private final Set<T> finalMarking;
    private final List<K> allModelMovables;
    private final List<T> allTokenables;
    private final List<F> extraMovables = new LinkedList<F>();
    private final GenerationDescription description;
    public BaseGenerationHelper(Collection<T> initialMarking, Collection<T> finalMarking, Collection<K> allModelMovables, Collection<T> allTokenables, GenerationDescription description)
    {
        this.initialMarking = new HashSet<T>(initialMarking);
        this.finalMarking = new HashSet<T>(finalMarking);
        this.allModelMovables = new ArrayList<K>(allModelMovables);
        this.allTokenables = new ArrayList<T>(allTokenables);
        this.description = description;
        moveToInitialState();
    }

    @Override
    public GenerationDescription getGenerationDescription()
    {
        return description;
    }

    protected abstract void putInitialToken(T tokenable);

    @Override
    public List<K> getAllModelMovables()
    {
        return Collections.unmodifiableList(allModelMovables);
    }


    @Override
    public List<F> getExtraMovables()
    {
        return Collections.unmodifiableList(extraMovables);
    }

    @Override
    public void moveToInitialState()
    {
        removeAllTokens();
        extraMovables.clear();
        for (T initialTokenable : initialMarking)
        {
            putInitialToken(initialTokenable);
        }
    }

    protected void removeAllTokens()
    {
        for (Tokenable tokenable : allTokenables)
        {
            tokenable.removeAllTokens();
        }
    }

    protected <L extends Movable> L pickRandomMovable(List<L> movables)
    {
        int index = random.nextInt(movables.size());
        return movables.get(index);
    }

    @Override
    @SuppressWarnings("unchecked")
    public AssessedMovementResult handleMovementResult(MovementResult movementResult)
    {
        if (movementResult == null)
        {
            throw new IllegalArgumentException("Movement result cannot be null");
        }

        extraMovables.removeAll(movementResult.getConsumedExtraMovables());
        extraMovables.addAll(movementResult.getProducedExtraMovables());

        boolean replayCompleted = tokensOnlyInFinalMarking();
        return new AssessedMovementResult(replayCompleted, true);
    }

    //returns true if final marking was reached
    protected boolean tokensOnlyInFinalMarking()
    {
        for (Tokenable tokenable : allTokenables)
        {
            if (tokenable.hasTokens())
            {
                if (!finalMarking.contains(tokenable))
                {
                    return false;
                }
            }
        }

        return true;
    }
}

File: src/main/java/org/processmining/utils/helpers/BaseGenerationHelper.kt
package org.processmining.utils.helpers

import org.processmining.models.*

import java.util.*

/**
 * Created by Ivan Shugurov on 23.10.2014.
 */

/**
 * @param <T> type of objects which can contain tokens, for example, places in Petri nets
 * @param <K> type of objects it's possible to move through. For example, transitions in Petri nets
 * @param <F> type of additional movables possible during a replay
</F></K></T> */
abstract class BaseGenerationHelper<T : Tokenable<*>, K : Movable, F : Movable>(
        initialMarking: Collection<T>,
        finalMarking: Collection<T>,
        allModelMovables: Collection<K>,
        allTokenables: Collection<T>,
        override val generationDescription: GenerationDescription
) : GenerationHelper<K, F> {
    
    // everything is immutable
    
    private val initialMarking = initialMarking
    private val finalMarking = finalMarking
    
    private val allTokenables = allTokenables.toList()
    
    override val allModelMovables = allModelMovables.toList()
    override val extraMovables = LinkedList<F>()
    
    init {
        moveToInitialState()
    }
    
//    override fun getGenerationDescription(): GenerationDescription {
//        return description
//    }
    
    protected abstract fun putInitialToken(place: T)
    
    
    override fun moveToInitialState() {
        removeAllTokens()
        extraMovables.clear()
        for (initialTokenable in initialMarking) {
            putInitialToken(initialTokenable)
        }
    }
    
    protected fun removeAllTokens() {
        for (tokenable in allTokenables) {
            tokenable.removeAllTokens()
        }
    }
    
    protected fun <L : Movable> pickRandomMovable(movables: List<L>): L {
        val index = random.nextInt(movables.size)
        return movables[index]
    }
    
    
    override fun handleMovementResult(movementResult: MovementResult<F>): AssessedMovementResult {
        
        extraMovables.removeAll(movementResult.consumedExtraMovables)
        extraMovables.addAll(movementResult.producedExtraMovables)
        
        val replayCompleted = tokensOnlyInFinalMarking()
        return AssessedMovementResult(replayCompleted, true)
    }
    
    //returns true if final marking was reached
    protected open fun tokensOnlyInFinalMarking(): Boolean {
        return allTokenables.filter { it.hasTokens() }.all { finalMarking.contains(it) }
    }
    
    companion object {
        @JvmStatic
        protected val random = Random()
    }
}